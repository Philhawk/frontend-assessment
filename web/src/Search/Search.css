.search-field {
	padding-left: 32px;
	height: 40px;
	width: 70%;
	display: block;
	position: relative;
	margin-left: auto;
	margin-right: auto;
	border: solid 1px grey;
	font-size: 16px;
	border-radius: 5px;
	font-weight: 500;
	margin-bottom: 12px;
}

.search-results {
	display: -webkit-flex;
	display: flex;
	-webkit-align-items: center;
	align-items: center;
	width: 285px;
	margin-left: auto;
	margin-right: auto;
	padding: 5px;
	border-bottom: 1px grey solid;
	border-left: 1px grey solid;
	border-right: 1px grey solid;
	background-color: #FAFAFA;
}

.spinner {
	height: 38px;
	margin-top: -54px;
	position: absolute;
	margin-right: 12%;
	right: 0;
}

.search-container {
	overflow-y: scroll;
	height: 230px;
}
.no-results {
	padding-left: 32px;
	height: 150px;
	width: 70%;
	display: block;
	position: relative;
	margin-left: auto;
	margin-right: auto;
	border: solid 1px grey;
	font-size: 16px;
	border-radius: 5px;
	background-color: #FAFAFA
}
.search-container > div:first-child {
	border-radius: 10px 10px 0 0;
	border: 1px grey solid;
}
.search-container > div:last-child {
	border-radius: 0 0 10px 10px;
}
