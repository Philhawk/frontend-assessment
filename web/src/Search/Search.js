import React, { Component } from 'react'
import './Search.css';
import axios from 'axios'
import spinner from '../Assets/spinner.svg'
import SearchResults from '../SearchResults/SearchResults'
import NoResults from '../NoResults/NoResults'

class Search extends Component {
  constructor() {
    super();
    this.state = {
      query: '',
      results: [],
      showSpinner: false,
      showNoResults: false
    }

    this.selectHero = this.selectHero.bind(this);
  }

  selectHero = (hero) => {
    this.setState({
      results: [],
      showNoResults: false,
      showSpinner: false,
      query: hero
    })
  }

  getInfo = (query) => {
    console.log('ping')
    this.setState({showSpinner: false, results: []})
    axios.get(`http://localhost:1337/superheroes?name_like=${query}`)
      .then(({ data }) => {
        this.setState({
          results: data,      
          showSpinner: false                     
        })
      }).catch((err) => {
        // TODO
      })
  }

  handleInputChange = () => {
    let shouldSpinnerBeSet = this.state.results.length !== 1 ? true : false;
    this.setState({
      showSpinner: shouldSpinnerBeSet,
      query: this.search.value
    }, () => {
      if (this.state.query.length > 4 && this.state.results.length === 0) {
        this.setState({showNoResults: true, showSpinner: false})
      }
      else if (this.state.query && this.state.query.length > 1) {
        this.setState({showNoResults: false})
        this.getInfo(this.state.query)
      } else if (this.state.query === '') {
        this.setState({showSpinner: false, showNoResults: false})
      } else if (this.state.results.length === 0) {
        this.setState({results: [], showSpinner: false, showNoResults: true})
      } 
    })
  }

  render() {
    return (
      <form>
				<span>
					<input
						placeholder="Search.."
						ref={input => this.search = input}
						onChange={this.handleInputChange}
						className="search-field"
            value={this.state.query}
					/>
          <img style={{display: this.state.showNoResults ? '' : 'none'}} className="spinner" src={spinner}/>
				</span>
				<div style={{display: this.state.showNoResults ? 'none' : ''}} className="search-container">
        {
					this.state.results.length > 0 ? this.state.results.map((row, i) =>
					<div className="search-results">
						<SearchResults selectHero={this.selectHero} hero={row} />
					</div>
          ) : ''
        }
				</div>
        { this.state.showNoResults ?
          <div className="no-results"> 
            <NoResults query={this.state.query}/> 
          </div> : ''
        }
			</form>
    )
  }
}

export default Search