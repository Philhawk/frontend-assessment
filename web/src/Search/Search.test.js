import React from "react";
import { shallow } from "enzyme";
import Search from './Search';

it('Search Component snapshot test', () => {
    const wrapper = shallow(<Search />);
    expect(wrapper).toMatchSnapshot();
  });