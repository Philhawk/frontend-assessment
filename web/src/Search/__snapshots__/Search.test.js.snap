// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component snapshot test 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Search />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <span>
          <input
            className="search-field"
            onChange={[Function]}
            placeholder="Search by name/location.."
            value=""
          />
          <img
            className="spinner"
            src="spinner.svg"
            style={
              Object {
                "display": "none",
              }
            }
          />
        </span>,
        <div
          className="search-container"
          style={
            Object {
              "display": "",
            }
          }
        >
          
        </div>,
        "",
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="search-field"
              onChange={[Function]}
              placeholder="Search by name/location.."
              value=""
            />,
            <img
              className="spinner"
              src="spinner.svg"
              style={
                Object {
                  "display": "none",
                }
              }
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "search-field",
              "onChange": [Function],
              "placeholder": "Search by name/location..",
              "value": "",
            },
            "ref": [Function],
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "spinner",
              "src": "spinner.svg",
              "style": Object {
                "display": "none",
              },
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
        ],
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "",
          "className": "search-container",
          "style": Object {
            "display": "",
          },
        },
        "ref": null,
        "rendered": "",
        "type": "div",
      },
      "",
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <span>
            <input
              className="search-field"
              onChange={[Function]}
              placeholder="Search by name/location.."
              value=""
            />
            <img
              className="spinner"
              src="spinner.svg"
              style={
                Object {
                  "display": "none",
                }
              }
            />
          </span>,
          <div
            className="search-container"
            style={
              Object {
                "display": "",
              }
            }
          >
            
          </div>,
          "",
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="search-field"
                onChange={[Function]}
                placeholder="Search by name/location.."
                value=""
              />,
              <img
                className="spinner"
                src="spinner.svg"
                style={
                  Object {
                    "display": "none",
                  }
                }
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "search-field",
                "onChange": [Function],
                "placeholder": "Search by name/location..",
                "value": "",
              },
              "ref": [Function],
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "spinner",
                "src": "spinner.svg",
                "style": Object {
                  "display": "none",
                },
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
          ],
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "",
            "className": "search-container",
            "style": Object {
              "display": "",
            },
          },
          "ref": null,
          "rendered": "",
          "type": "div",
        },
        "",
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
